Avoid panic when core schema is nil

While core schema should always be set by the caller to a non-nil value (because it can always fall back to universal schema), the function should _not_ panic when this isn't the case, but probably just error with a human-readable message.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14e07ca]

goroutine 5 [running]:
github.com/hashicorp/terraform-schema/schema.MergeCoreWithJsonProviderSchemas(0xc0003bcc90, 0x0, 0xc00000eca0, 0x0, 0x0, 0x0)
	/Users/radeksimko/gopath/pkg/mod/github.com/hashicorp/terraform-schema@v0.0.0-20201110191417-e2e5d08913c4/schema/schema_merge.go:19 +0x3a
github.com/hashicorp/terraform-ls/internal/terraform/rootmodule.(*rootModule).MergedSchema(0xc000001800, 0x0, 0x0, 0x0)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform-ls/internal/terraform/rootmodule/root_module.go:527 +0x18e
github.com/hashicorp/terraform-ls/internal/terraform/rootmodule.(*rootModuleManager).SchemaForPath(0xc00011c880, 0xc00002e287, 0x28, 0x1864340, 0xc000001800, 0x0)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform-ls/internal/terraform/rootmodule/root_module_manager.go:151 +0x15f
github.com/hashicorp/terraform-ls/langserver/handlers.(*logHandler).TextDocumentComplete(0xc0000a86d8, 0x185c360, 0xc0003bd050, 0xc00002e640, 0x37, 0x12, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform-ls/langserver/handlers/complete.go:39 +0x255
reflect.Value.call(0x15ba100, 0xc000058490, 0x13, 0x1665aa2, 0x4, 0xc0003bd2c0, 0x2, 0x2, 0x18, 0xc00000ec00, ...)
	/Users/radeksimko/.goenv/versions/1.14.9/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0x15ba100, 0xc000058490, 0x13, 0xc0003bd2c0, 0x2, 0x2, 0x2, 0xc0003bd2c0, 0x1)
	/Users/radeksimko/.goenv/versions/1.14.9/src/reflect/value.go:321 +0xb4
github.com/creachadair/jrpc2/handler.newHandler.func8(0x185c360, 0xc0003bd050, 0xc0003ba380, 0x0, 0x0, 0x30, 0x1621f00)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/handler/handler.go:232 +0x20a
github.com/creachadair/jrpc2/handler.Func.Handle(...)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/handler/handler.go:23
github.com/hashicorp/terraform-ls/langserver/handlers.handle(0x185c360, 0xc0003bd050, 0xc0003ba380, 0x15ba100, 0xc000058490, 0xc00011c880, 0x185c360, 0xc0003bd050, 0x15fec80)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform-ls/langserver/handlers/service.go:345 +0x83
github.com/hashicorp/terraform-ls/langserver/handlers.(*service).Assigner.func10(0x185c360, 0xc0003bcf60, 0xc0003ba380, 0x0, 0x0, 0x0, 0x0)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform-ls/langserver/handlers/service.go:226 +0x203
reflect.Value.call(0x15d6140, 0xc00017d140, 0x13, 0x1665aa2, 0x4, 0xc0003bcfc0, 0x2, 0x2, 0x18, 0xc00000e9c0, ...)
	/Users/radeksimko/.goenv/versions/1.14.9/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0x15d6140, 0xc00017d140, 0x13, 0xc0003bcfc0, 0x2, 0x2, 0x2, 0xc0003bcfc0, 0x1)
	/Users/radeksimko/.goenv/versions/1.14.9/src/reflect/value.go:321 +0xb4
github.com/creachadair/jrpc2/handler.newHandler.func8(0x185c360, 0xc0003bcf60, 0xc0003ba380, 0x158c0c0, 0xc000058480, 0x15b9080, 0xc00000e980)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/handler/handler.go:232 +0x20a
github.com/creachadair/jrpc2/handler.Func.Handle(0xc00027c3c0, 0x185c360, 0xc0003bcf60, 0xc0003ba380, 0x0, 0x0, 0x185c360, 0xc0003bcf60)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/handler/handler.go:23 +0x44
github.com/creachadair/jrpc2.(*Server).invoke(0xc0002b8000, 0x185c2a0, 0xc0003ba3c0, 0x18508a0, 0xc00027c3c0, 0xc0003ba380, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/server.go:339 +0x19b
github.com/creachadair/jrpc2.(*Server).dispatch.func1.1()
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/server.go:225 +0xb7
github.com/creachadair/jrpc2.(*Server).dispatch.func1(0xc0002abfc8, 0x0)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/server.go:230 +0x161
github.com/creachadair/jrpc2.(*Server).serve.func1(0xc0002b8000, 0xc0004d1da0)
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/server.go:151 +0x51
created by github.com/creachadair/jrpc2.(*Server).serve
	/Users/radeksimko/gopath/pkg/mod/github.com/creachadair/jrpc2@v0.10.1/server.go:149 +0x6d
```

--- a/schema/errors.go
+++ b/schema/errors.go
@@ -1,0 +1,7 @@
+package schema
+
+type coreSchemaRequiredErr struct{}
+
+func (e coreSchemaRequiredErr) Error() string {
+	return "core schema required (none provided)"
+}

--- a/schema/schema_merge.go
+++ b/schema/schema_merge.go
@@ -1,6 +1,9 @@
 package schema
 
 import (
+	"fmt"
+
+	"github.com/hashicorp/go-version"
 	"github.com/hashicorp/hcl-lang/lang"
 	"github.com/hashicorp/hcl-lang/schema"
 	"github.com/hashicorp/hcl/v2"
@@ -12,21 +15,51 @@ import (
 type SchemaMerger struct {
 	coreSchema  *schema.BodySchema
 	parsedFiles map[string]*hcl.File
+
+	coreVersion      *version.Version
+	providerVersions map[addrs.Provider]*version.Version
 }
 
 func NewSchemaMerger(coreSchema *schema.BodySchema) *SchemaMerger {
 	return &SchemaMerger{
-		coreSchema:  coreSchema,
-		parsedFiles: make(map[string]*hcl.File, 0),
+		coreSchema:       coreSchema,
+		parsedFiles:      make(map[string]*hcl.File, 0),
+		providerVersions: make(map[addrs.Provider]*version.Version, 0),
 	}
 }
 
+// SetParsedFiles sets a map of parsed files where key is a filename
 func (m *SchemaMerger) SetParsedFiles(files map[string]*hcl.File) {
 	m.parsedFiles = files
 }
 
-// MergeWithJsonProviderSchemas provides a merged schema based on provided
-// parsed files, core schema and terraform-json formatted provider schema
+// SetCoreVersion sets version of Terraform (core) to help identify core schema
+// and schema of the builtin terraform provider
+func (m *SchemaMerger) SetCoreVersion(v *version.Version) {
+	m.coreVersion = v
+}
+
+// SetProviderVersions sets versions of providers to help identify
+// where the provider schemas came from
+func (m *SchemaMerger) SetProviderVersions(versions map[string]*version.Version) error {
+	versionMap := make(map[addrs.Provider]*version.Version, 0)
+
+	for addr, ver := range versions {
+		srcAddr, err := addrs.ParseProviderSourceString(addr)
+		if err != nil {
+			return err
+		}
+		versionMap[srcAddr] = ver
+	}
+
+	m.providerVersions = versionMap
+
+	return nil
+}
+
+// MergeWithJsonProviderSchemas provides a merged schema based on
+// terraform-json formatted provider schema and any other data
+// provided via setters
 func (m *SchemaMerger) MergeWithJsonProviderSchemas(ps *tfjson.ProviderSchemas) (*schema.BodySchema, error) {
 	if m.coreSchema == nil {
 		return nil, coreSchemaRequiredErr{}
@@ -73,10 +106,7 @@ func (m *SchemaMerger) MergeWithJsonProviderSchemas(ps *tfjson.ProviderSchemas) 
 			providerSchema = provider.ConfigSchema.Block
 		}
 
-		detail := srcAddr.ForDisplay()
-		if srcAddr.IsBuiltIn() {
-			detail = "(builtin)"
-		}
+		detail := m.detailForSrcAddr(srcAddr)
 
 		for _, localRef := range localRefs {
 			mergedSchema.Blocks["provider"].DependentBody[schema.NewSchemaKey(schema.DependencyKeys{
@@ -134,6 +164,25 @@ func (m *SchemaMerger) MergeWithJsonProviderSchemas(ps *tfjson.ProviderSchemas) 
 	}
 
 	return mergedSchema, nil
+}
+
+func (m *SchemaMerger) detailForSrcAddr(addr addrs.Provider) string {
+	if addr.IsBuiltIn() {
+		if m.coreVersion == nil {
+			return "(builtin)"
+		}
+		return fmt.Sprintf("(builtin %s)", m.coreVersion.String())
+	}
+
+	detail := addr.ForDisplay()
+	for pAddr, ver := range m.providerVersions {
+		if addr.Equals(pAddr) {
+			detail += " " + ver.String()
+			break
+		}
+	}
+
+	return detail
 }
 
 func convertBodySchemaFromJson(detail string, schemaBlock *tfjson.SchemaBlock) *schema.BodySchema {

--- a/schema/schema_merge.go
+++ b/schema/schema_merge.go
@@ -9,12 +9,34 @@ import (
 	"github.com/hashicorp/terraform-schema/internal/refdecoder"
 )
 
-// MergeCoreWithJsonProviderSchemas provides a merged schema based on provided
-// parsed files, core schema and terraform-json formatted provider schema
-func MergeCoreWithJsonProviderSchemas(m map[string]*hcl.File, coreSchema *schema.BodySchema, ps *tfjson.ProviderSchemas) (
-	*schema.BodySchema, error) {
+type SchemaMerger struct {
+	coreSchema  *schema.BodySchema
+	parsedFiles map[string]*hcl.File
+}
 
-	mergedSchema := coreSchema
+func NewSchemaMerger(coreSchema *schema.BodySchema) *SchemaMerger {
+	return &SchemaMerger{
+		coreSchema:  coreSchema,
+		parsedFiles: make(map[string]*hcl.File, 0),
+	}
+}
+
+func (m *SchemaMerger) SetParsedFiles(files map[string]*hcl.File) {
+	m.parsedFiles = files
+}
+
+// MergeWithJsonProviderSchemas provides a merged schema based on provided
+// parsed files, core schema and terraform-json formatted provider schema
+func (m *SchemaMerger) MergeWithJsonProviderSchemas(ps *tfjson.ProviderSchemas) (*schema.BodySchema, error) {
+	if m.coreSchema == nil {
+		return nil, coreSchemaRequiredErr{}
+	}
+
+	if ps == nil {
+		return m.coreSchema, nil
+	}
+
+	mergedSchema := m.coreSchema
 
 	if mergedSchema.Blocks["provider"].DependentBody == nil {
 		mergedSchema.Blocks["provider"].DependentBody = make(map[schema.SchemaKey]*schema.BodySchema)
@@ -26,15 +48,15 @@ func MergeCoreWithJsonProviderSchemas(m map[string]*hcl.File, coreSchema *schema
 		mergedSchema.Blocks["data"].DependentBody = make(map[schema.SchemaKey]*schema.BodySchema)
 	}
 
-	refs, err := refdecoder.DecodeProviderReferences(m)
+	refs, err := refdecoder.DecodeProviderReferences(m.parsedFiles)
 	if err != nil {
-		return coreSchema, err
+		return m.coreSchema, err
 	}
 
 	for sourceString, provider := range ps.Schemas {
 		srcAddr, err := addrs.ParseProviderSourceString(sourceString)
 		if err != nil {
-			return coreSchema, err
+			return m.coreSchema, err
 		}
 
 		localRefs := refs.LocalNamesByAddr(srcAddr)

--- a/schema/schema_merge_test.go
+++ b/schema/schema_merge_test.go
@@ -2,6 +2,7 @@ package schema
 
 import (
 	"encoding/json"
+	"errors"
 	"io/ioutil"
 	"testing"
 
@@ -14,7 +15,29 @@ import (
 	"github.com/zclconf/go-cty/cty"
 )
 
-func TestMergeCoreWithJsonProviderSchemas_v012(t *testing.T) {
+func TestMergeWithJsonProviderSchemas_noCoreSchema(t *testing.T) {
+	sm := NewSchemaMerger(nil)
+
+	_, err := sm.MergeWithJsonProviderSchemas(nil)
+	if err == nil {
+		t.Fatal("expected error for nil core schema")
+	}
+
+	if !errors.Is(err, coreSchemaRequiredErr{}) {
+		t.Fatalf("unexpected error: %#v", err)
+	}
+}
+
+func TestMergeWithJsonProviderSchemas_noProviderSchema(t *testing.T) {
+	sm := NewSchemaMerger(testCoreSchema)
+
+	_, err := sm.MergeWithJsonProviderSchemas(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMergeWithJsonProviderSchemas_v012(t *testing.T) {
 	b, err := ioutil.ReadFile("testdata/test-config-0.12.tf")
 	if err != nil {
 		t.Fatal(err)
@@ -34,11 +57,12 @@ func TestMergeCoreWithJsonProviderSchemas_v012(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	files := map[string]*hcl.File{
+	sm := NewSchemaMerger(testCoreSchema)
+	sm.SetParsedFiles(map[string]*hcl.File{
 		"test.tf": f,
-	}
+	})
 
-	mergedSchema, err := MergeCoreWithJsonProviderSchemas(files, testCoreSchema, ps)
+	mergedSchema, err := sm.MergeWithJsonProviderSchemas(ps)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -52,7 +76,7 @@ func TestMergeCoreWithJsonProviderSchemas_v012(t *testing.T) {
 	}
 }
 
-func TestMergeCoreWithJsonProviderSchemas_v013(t *testing.T) {
+func TestMergeWithJsonProviderSchemas_v013(t *testing.T) {
 	b, err := ioutil.ReadFile("testdata/test-config-0.13.tf")
 	if err != nil {
 		t.Fatal(err)
@@ -72,11 +96,12 @@ func TestMergeCoreWithJsonProviderSchemas_v013(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	files := map[string]*hcl.File{
+	sm := NewSchemaMerger(testCoreSchema)
+	sm.SetParsedFiles(map[string]*hcl.File{
 		"test.tf": f,
-	}
+	})
 
-	mergedSchema, err := MergeCoreWithJsonProviderSchemas(files, testCoreSchema, ps)
+	mergedSchema, err := sm.MergeWithJsonProviderSchemas(ps)
 	if err != nil {
 		t.Fatal(err)
 	}

